src/main/java/array/Array2d.java

```diff
@@ -19,31 +36,73 @@
      * @return array of the given number of rows and columns
      */
     public static <T> Array2d<T> of(int numberOfRows, int numberOfColumns) {
-        throw new UnsupportedOperationException();
+        return new Array2d<>(numberOfRows, numberOfColumns);
     }
 
     @Override
     public T insert(Coordinates2d coordinates, T value) {
-        throw new UnsupportedOperationException();
+        validateIndex(coordinates);
+        if (array[coordinates.getRow()][coordinates.getColumn()] != null) {
+            throw new IllegalArgumentException("this cell is already occupied");
+        }
+        array[coordinates.getRow()][coordinates.getColumn()] = value;
+        return value;
     }
 
     @Override
     public Optional<T> at(Coordinates2d coordinates) {
-        throw new UnsupportedOperationException();
+        validateIndex(coordinates);
+        T value = (T) array[coordinates.getRow()][coordinates.getColumn()];
+        if (value == null) {
+            return Optional.empty();
+        } else {
+            return Optional.of(value);
+        }
     }
 
     @Override
     public Optional<Coordinates2d> search(T value) {
-        throw new UnsupportedOperationException();
+        validateValue(value);
+        for (int i = 0; i < numberOfRows; i++) {
+            for (int j = 0; j < numberOfColumns; j++) {
+                if (array[i][j] != null && array[i][j].equals(value)) {
```

Comment:
```suggestion
                if (value.equals(array[i][j])) {
```
'cause value is already not null

Reply:
fixed